src/backend/common/Buffer.h

```diff
@@ -34,7 +34,7 @@ namespace backend {
             bool IsTransitionPossible(nxt::BufferUsageBit usage) const;
             bool IsFrozen() const;
             bool HasFrozenUsage(nxt::BufferUsageBit usage) const;
-            void TransitionUsageImpl(nxt::BufferUsageBit usage);
+            virtual void TransitionUsageImpl(nxt::BufferUsageBit usage);
```

Comment:
In a patch before the D3D12 part, let's change TransitionUsageImpl do ```virtual void TransitionUsageImpl(currentUsage, targetUsage) = 0```. TransitionUsageImpl is used in a couple places: 

 - ```BufferBase::TransitionUsage``` to change usage, this would instead call the Impl and then set the current usage.
 - Backend command buffers, this is because if there is a transition in the command buffers, this should be reflected in the current usage after queue->execute. You can introduce a ```UpdateUsageInternal``` to buffer that will just asser TranstitionPossible and update the member.
 - FreezeUsage (in this PR). It would do the same as in the first bullet.

The same change should be reflected in TextureBase.